src/hooks/mocha.ts

```diff
@@ -0,0 +1,66 @@
+import { simple as walk } from "acorn-walk";
+import type { ESTree } from "meriyah";
+import { expressionFor } from ".";
+import { call_, identifier } from "../generate";
+import { EventEmitter } from "node:stream";
+import Recording from "../Recording";
+import { recording, start } from "../recorder";
+
+export function shouldInstrument(url: URL): boolean {
+  return url.pathname.endsWith("/mocha/lib/runner.js");
+}
+
+export function transform(program: ESTree.Program): ESTree.Program {
+  return patchRuntime(program);
+}
+
+const EVENT_TEST_BEGIN = 'test';
+const EVENT_TEST_PASS = 'pass';
+const EVENT_TEST_FAIL = 'fail';
+const EVENT_TEST_END = 'test end';
+
+interface Test {
+  title: string
+  fullTitle: () => string
+}
+
+function registerEventListeners(runner: EventEmitter) {
+  runner.on(EVENT_TEST_BEGIN, function (test: Test) {
+    const rec = new Recording("tests", "mocha", test.fullTitle());
+    start(rec);
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_PASS, function (test: Test) {
+    recording.metadata.test_status = "succeeded";
+    recording.finish();
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_FAIL, function (test: Test) {
+    recording.metadata.test_status = "failed";
+    recording.finish();
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_END, function (test: Test) {
+    ;
+  });
+}
+
+export function patchRuntime(program: ESTree.Program): ESTree.Program {
+  recording.abandon();
+  // Find constructor of mocha Runner in ".../mocha/lib/runner.js".
+  // Append an expression to constructor body for registering test
+  // event listeners.
+  walk(program, {
+    MethodDefinition(md: ESTree.MethodDefinition) {
+      if (md.kind === "constructor") {
+        md.value.body?.body.push({
+          type: "ExpressionStatement",
+          expression: call_(expressionFor(registerEventListeners), identifier("this"))})
```

Comment:
BTW, there's `this_` in `generate` that can be used instead of `identifier("this")`.